larpandoracontent/LArThreeDReco/LArHitCreation/RANSACMethod.cc

```diff
@@ -0,0 +1,490 @@
+
+    hit.SetDisplacement(displacement);
+}
+
+void hitDisplacement(RANSACHit &hit, const CartesianVector fitEnd, const CartesianVector fitDirection)
+{
+    const CartesianVector hitPosition = hit.GetProtoHit().GetPosition3D();
+    const float displacement = (hitPosition - fitEnd).GetCrossProduct(fitDirection).GetMagnitude();
+
+    hit.SetDisplacement(displacement);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArRANSACMethod::ExtendFit(
+    std::list<RANSACHit> &hitsToTestAgainst,
+    RANSACHitVector &hitsToUseForFit,
+    std::vector<RANSACHit> &hitsToAdd,
+    const float distanceToFitThreshold,
+    const ExtendDirection extendDirection
+)
+{
+    if (hitsToUseForFit.size() == 0)
+        return;
+
+    float distanceToEndThreshold = 10; // TODO: Config?
+
+    CartesianPointVector fitPoints;
+    for (auto hit : hitsToUseForFit)
+        fitPoints.push_back(hit.GetProtoHit().GetPosition3D());
+
+    const unsigned int layerWindow(100);
+    const float pitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    const ThreeDSlidingFitResult slidingFit(&fitPoints, layerWindow, pitch);
+
+    CartesianVector fitDirection = slidingFit.GetGlobalMaxLayerDirection();
+    CartesianVector fitEnd = slidingFit.GetGlobalMaxLayerPosition();
+
+    if (extendDirection == ExtendDirection::Backward)
+    {
+        // TODO: This seems a bit iffy...does this work as I expect?
+        fitDirection = slidingFit.GetGlobalMinLayerDirection();
+        fitEnd = slidingFit.GetGlobalMinLayerPosition();
+        fitDirection = fitDirection * -1.0;
+    }
+
+    // INFO: Check if there is a detector gap, and if there is, extend the fit over it.
+    auto distanceToProjectOverGap = LArGeometryHelper::ProjectAcrossGap3D(this->GetPandora(), fitEnd, fitDirection, 2.0, 10);
+    if (distanceToProjectOverGap > 0)
+        distanceToEndThreshold += std::min(distanceToProjectOverGap, 50.0f);
+
+    // ATTN: This is done in 3 stages to split up the 3 different qualities of
+    //       hits:
+    //          - Hits based on the fit projections.
+    //          - Hits based against the fit.
+    //          - Unfavourable hits.
+    auto projectedDisplacementTest = [&slidingFit] (RANSACHit &hit, float threshold) {
+        projectedHitDisplacement(hit, slidingFit);
+        if (hit.GetDisplacement() < threshold && hit.IsFavourable())
+            return true;
+        return false;
+    };
+    auto displacementTest = [&fitEnd, &fitDirection] (RANSACHit &hit, float threshold) {
+        hitDisplacement(hit, fitEnd, fitDirection);
+        if (hit.GetDisplacement() < threshold && hit.IsFavourable())
+            return true;
+        return false;
+    };
+    auto unfavourableTest = [] (RANSACHit &hit, float threshold) {
+        if (hit.GetDisplacement() < threshold)
+            return true;
+        return false;
+    };
```

Comment:
Can make these a little more concise by just returning the result of the conditional directly.